File: .claude/hooks/task-completed-test-gate.mjs
#!/usr/bin/env node
// @hookstack task-completed-test-gate
import { execSync } from "node:child_process";
// @hookstack task-completed-test-gate
// Bloque la complétion d'une tâche si les tests échouent (TaskCompleted)
import { existsSync, readFileSync } from "node:fs";
import { join } from "node:path";
import { fileURLToPath } from "node:url";

function defaultExec(cmd) {
	return execSync(cmd, { stdio: "pipe", timeout: 120_000 });
}

// Détecte le gestionnaire de paquets depuis le lockfile (cohérent avec enforce-package-managers).
function detectManager({
	exists = existsSync,
	projectDir = process.cwd(),
} = {}) {
	if (exists(join(projectDir, "pnpm-lock.yaml"))) return "pnpm";
	if (
		exists(join(projectDir, "bun.lockb")) ||
		exists(join(projectDir, "bun.lock"))
	)
		return "bun";
	if (exists(join(projectDir, "yarn.lock"))) return "yarn";
	return "npm";
}

const PY_MARKERS = ["pyproject.toml", "setup.py", "setup.cfg", "pytest.ini"];

// Le gate Python ne s'arme que si le projet déclare réellement des tests
// (pytest.ini explicite ou dossier tests/) — sinon `uv run pytest` exit 5
// (no tests ran) bloquerait la complétion de chaque tâche sur un projet sans
// test. pyproject.toml seul ne suffit pas : tout projet uv en possède un.
function hasPythonTests({ exists, projectDir }) {
	if (!PY_MARKERS.some((f) => exists(join(projectDir, f)))) return false;
	if (exists(join(projectDir, "pytest.ini"))) return true;
	return exists(join(projectDir, "tests")) || exists(join(projectDir, "test"));
}

export function run(
	input,
	{
		exec = defaultExec,
		exists = existsSync,
		projectDir = process.env.CLAUDE_PROJECT_DIR ?? process.cwd(),
	} = {},
) {
	const commands = [];
	// Python : gate pytest via uv (même outil que stop-pytest).
	if (hasPythonTests({ exists, projectDir })) commands.push("uv run pytest -q");
	// Node : gate sur le gestionnaire détecté, `--if-present` pour sauter sans script test.
	if (exists(join(projectDir, "package.json")))
		commands.push(`${detectManager({ exists, projectDir })} test --if-present 2>&1`);
	try {
		for (const cmd of commands) exec(cmd);
		return null;
	} catch (e) {
		const out = (e.stdout ?? e.stderr ?? e.message).toString().slice(0, 800);
		return {
			exitCode: 2,
			message: `Tests must pass before completing "${input.task_subject}".\n${out}`,
		};
	}
}

/* v8 ignore next 6 */
if (process.argv[1] === fileURLToPath(import.meta.url)) {
	const input = JSON.parse(readFileSync(0, "utf8"));
	const result = run(input);
	if (result) {
		process.stderr.write(result.message);
		process.exit(result.exitCode);
	}
}

File: .claude/hooks/task-completed-test-gate.py
#!/usr/bin/env python3
# @hookstack task-completed-test-gate
"""Blocks task completion if the pytest suite fails (TaskCompleted). Python-only variant."""
import json
import os
import subprocess
import sys

PY_MARKERS = ["pyproject.toml", "setup.py", "setup.cfg", "pytest.ini"]


def _exec(cmd, timeout=120):
    return subprocess.run(
        cmd, shell=True, capture_output=True, text=True, timeout=timeout, check=True
    )


# Le gate Python ne s'arme que si le projet déclare réellement des tests
# (pytest.ini explicite ou dossier tests/) — sinon `uv run pytest` exit 5
# (no tests ran) bloquerait la complétion de chaque tâche sur un projet sans
# test. pyproject.toml seul ne suffit pas : tout projet uv en possède un.
def _has_python_tests(exists, project_dir):
    if not any(exists(f) for f in PY_MARKERS):
        return False
    if exists("pytest.ini"):
        return True
    return exists("tests") or exists("test")


def run(input_data, *, exec_cmd=None, exists=None, project_dir=None):
    if project_dir is None:
        project_dir = os.environ.get("CLAUDE_PROJECT_DIR") or os.getcwd()
    if exists is None:
        exists = lambda f: os.path.exists(os.path.join(project_dir, f))
    if exec_cmd is None:
        exec_cmd = _exec

    if not _has_python_tests(exists, project_dir):
        return None

    try:
        exec_cmd("uv run pytest -q")
    except subprocess.CalledProcessError as err:
        out = (err.stdout or err.stderr or str(err))[:800]
        subject = (input_data or {}).get("task_subject", "")
        return {
            "exitCode": 2,
            "message": f'Tests must pass before completing "{subject}".\n{out}',
        }
    return None


if __name__ == "__main__":
    data = json.load(sys.stdin)
    result = run(data)
    if result:
        sys.stderr.write(result["message"])
        sys.exit(result["exitCode"])
